### PR TITLE
add typing

### PR DESCRIPTION
**Pull request by Assistest**
Python supports typing, but Scrapy does not
I don't know if my actions are correct, I haven't submitted a PR
Scrapy is the first time
Please explain any errors. Thank you

image
image
image
image
image
image

### CODE REVIEW

**Comment by wRAR** [comment]
> Python supports typing, but Scrapy does not

Scrapy definitely supports typing and already ships many type hints, it's just not fully covered by them yet.

I haven't reviewed your changes yet, but I see that the typing check fails, so you should review and fix the reported issues, running it locally.

File: scrapy/http/response/__init__.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def __init__(
         self.status = int(status)
         self._set_body(body)
         self._set_url(url)
-        self.request = request
+        self.request: Request = request
```
Comment (wRAR) [inline scrapy/http/response/__init__.py:57]: This is incorrect, `request` can be None.

File: scrapy/pipelines/media.py [hunk 2]
```diff
@@ -76,10 +81,10 @@ def from_crawler(cls, crawler):
         pipe._fingerprinter = crawler.request_fingerprinter
         return pipe
 
-    def open_spider(self, spider):
+    def open_spider(self, spider: Spider):
         self.spiderinfo = self.SpiderInfo(spider)
 
-    def process_item(self, item, spider):
+    def process_item(self, item: dict, spider: Spider):
```
Comment (wRAR) [inline scrapy/pipelines/media.py:87]: `item` is not necessarily a dict.

File: scrapy/pipelines/media.py [hunk 3]
```diff
@@ -189,23 +194,23 @@ def _cache_result_and_execute_waiters(self, result, fp, info):
             defer_result(result).chainDeferred(wad)
 
     # Overridable Interface
-    def media_to_download(self, request, info, *, item=None):
+    def media_to_download(self, request: Request, info: SpiderInfo, *, item: dict=None):
```
Comment (wRAR) [inline scrapy/pipelines/media.py:197]: `item` is not necessarily a dict but in this case it can also be None.

File: scrapy/pipelines/media.py [hunk 4]
```diff
@@ -189,23 +194,23 @@ def _cache_result_and_execute_waiters(self, result, fp, info):
...
         """Check request before starting download"""
         pass
 
-    def get_media_requests(self, item, info):
+    def get_media_requests(self, item: dict, info: SpiderInfo):
         """Returns the media requests to download"""
         pass
 
-    def media_downloaded(self, response, request, info, *, item=None):
+    def media_downloaded(self, response: HtmlResponse, request: Request, info: SpiderInfo, *, item: dict=None):
```
Comment (wRAR) [inline scrapy/pipelines/media.py:205]: `response` is not necessarily a `HtmlResponse`.

File: scrapy/pipelines/media.py [hunk 5]
```diff
@@ -189,23 +194,23 @@ def _cache_result_and_execute_waiters(self, result, fp, info):
...
         """Handler for success downloads"""
         return response
 
     def media_failed(self, failure, request, info):
         """Handler for failed downloads"""
         return failure
 
-    def item_completed(self, results, item, info):
+    def item_completed(self, results: List[Tuple], item: dict, info: SpiderInfo):
```
Comment (wRAR) [inline scrapy/pipelines/media.py:213]: You should specify the tuple members types here.

File: scrapy/pipelines/media.py [hunk 6]
```diff
@@ -218,6 +223,6 @@ def item_completed(self, results, item, info):
                     )
         return item
 
-    def file_path(self, request, response=None, info=None, *, item=None):
+    def file_path(self, request: Request, response: HtmlResponse=None, info: SpiderInfo=None, *, item: dict=None):
```
Comment (wRAR) [inline scrapy/pipelines/media.py:226]: You should read about `Optional`.

File: scrapy/templates/project/module/middlewares.py.tmpl [hunk 7]
```diff
@@ -15,35 +20,35 @@ class ${ProjectName}SpiderMiddleware:
...
-    def process_spider_exception(self, response, exception, spider):
+    def process_spider_exception(self, response: HtmlResponse, exception, spider: Spider):
         # Called when a spider or process_spider_input() method
         # (from other spider middleware) raises an exception.
 
         # Should return either None or an iterable of Request or item objects.
         pass
 
-    def process_start_requests(self, start_requests, spider):
+    def process_start_requests(self, start_requests: Generator[Request], spider: Spider):
```
Comment (wRAR) [inline scrapy/templates/project/module/middlewares.py.tmpl:51]: `start_requests` is not necessarily a `Generator[Request]`.

File: scrapy/templates/project/module/pipelines.py.tmpl [hunk 8]
```diff
@@ -6,8 +6,10 @@
 
 # useful for handling different item types with a single interface
 from itemadapter import ItemAdapter
+from scrapy.spiders import Spider
+from $project_name.items import ${ProjectName}Item
 
 
 class ${ProjectName}Pipeline:
-    def process_item(self, item, spider):
+    def process_item(self, item: ${ProjectName}Item, spider: Spider):
```
Comment (wRAR) [inline scrapy/templates/project/module/pipelines.py.tmpl:14]: item is not necessarily a `${ProjectName}Item`

**Review by wRAR — CHANGES_REQUESTED** [review]

**Comment by Assistest** [comment]
I'm sorry I didn't understand enough  @wRAR

**Pull request closed without merging**